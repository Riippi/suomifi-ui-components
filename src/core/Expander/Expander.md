```jsx
import { Expander } from 'suomifi-ui-components';

<Expander title="Test expander" className="expander-test">
  Test expander
</Expander>;
```

```jsx
import { Expander } from 'suomifi-ui-components';

<Expander.group OpenAll="Open all" CloseAll="Close all">
  <Expander title="Test expander 1">Test expander content 1</Expander>
  <Expander title="Test expander 2">Test expander content 2</Expander>
  <Expander title="Test expander 3">Test expander content 3</Expander>
</Expander.group>;
```

## Controlled

- State for the individual Expanders are stored outside of the component and user has full control.
- Therefore when clicking the individual Expander they are not opened by default, user have to give the logic to change it.
- It's user's responsibility to keep the state stored outside to be updated as Open/Close All is used.
- `defaultOpen` prop will not work when Expander is in controlled state == `open` prop is given.

```jsx
import { Expander } from 'suomifi-ui-components';

const [expanderThreeOpen, setExpanderThreeOpen] = React.useState(
  false
);

<>
  <Expander.group OpenAll="Open all" CloseAll="Close all">
    <Expander title="Test expander 1">
      Test expander content 1
    </Expander>
    <Expander title="Test expander 2">
      Test expander content 2
    </Expander>
    <Expander
      title="Test expander 3"
      open={expanderThreeOpen}
      onClick={({ openState }) => {
        if (window.confirm('Toggle Expander 3')) {
          setExpanderThreeOpen(!openState);
        }
      }}
    >
      Test expander content 3
    </Expander>
  </Expander.group>
</>;
```
