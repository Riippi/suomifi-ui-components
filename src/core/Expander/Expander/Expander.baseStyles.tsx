import { css } from 'styled-components';
import { defaultThemeTokens as theme } from '../../theme';
import { element, font } from '../../theme/reset';

export const baseStyles = css`
  ${element({ theme })}
  ${font({ theme })('bodyText')}
    background-color: ${theme.colors.whiteBase};
  position: relative;
  padding: 0;
  border-radius: ${theme.radius.basic};
  box-shadow: ${theme.shadows.panelShadow};
  display: block;
  width: 100%;
  max-width: 100%;

  &:before {
    background-color: ${theme.colors.highlightLight4};
    opacity: 0;
  }
`;
