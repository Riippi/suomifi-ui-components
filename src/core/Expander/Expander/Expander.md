```jsx
import {
  Expander,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

<Expander className="expander-test" defaultOpen={true}>
  <ExpanderTitleButton>Test expander</ExpanderTitleButton>
  <ExpanderContent>Test expander</ExpanderContent>
</Expander>;
```

```jsx
import {
  Expander,
  ExpanderTitle,
  ExpanderContent,
  Checkbox
} from 'suomifi-ui-components';

<Expander className="expander-test" defaultOpen={true}>
  <ExpanderTitle>
    <Checkbox hintText="Checkbox hint text">Checkbox label</Checkbox>
  </ExpanderTitle>
  <ExpanderContent>Test expander</ExpanderContent>
</Expander>;
```

```jsx
import {
  Expander,
  ExpanderGroup,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

<ExpanderGroup
  OpenAllText="Open all"
  AriaOpenAllText="Open all expanders"
  CloseAllText="Close all"
  AriaCloseAllText="Close all expanders"
>
  <Expander>
    <ExpanderTitleButton>Test expander 1</ExpanderTitleButton>
    <ExpanderContent>Test expander content 1</ExpanderContent>
  </Expander>
  <Expander>
    <ExpanderTitleButton>Test expander 2</ExpanderTitleButton>
    <ExpanderContent>Test expander content 2</ExpanderContent>
  </Expander>
  <Expander>
    <ExpanderTitleButton>Test expander 3</ExpanderTitleButton>
    <ExpanderContent>Test expander content 3</ExpanderContent>
  </Expander>
</ExpanderGroup>;
```

## Controlled

- State for the individual Expanders are stored outside of the component and user has full control.
- Therefore when clicking the individual Expander they are not opened by default, user have to give the logic to change it.
- It's user's responsibility to keep the state stored outside to be updated as Open/Close All is used.
- `defaultOpen` prop will not work when Expander is in controlled state == `open` prop is given.

```jsx
import {
  Expander,
  ExpanderGroup,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

const [expanderThreeOpen, setExpanderThreeOpen] = React.useState(
  false
);

<>
  <ExpanderGroup
    OpenAllText="Open all"
    AriaOpenAllText="Open all expanders"
    CloseAllText="Close all"
    AriaCloseAllText="Close all expanders"
  >
    <Expander>
      <ExpanderTitleButton>Test expander 1</ExpanderTitleButton>
      <ExpanderContent>Test expander content 1</ExpanderContent>
    </Expander>
    <Expander>
      <ExpanderTitleButton>Test expander 2</ExpanderTitleButton>
      <ExpanderContent>Test expander content 2</ExpanderContent>
    </Expander>
    <Expander
      open={expanderThreeOpen}
      onOpenChange={(open) => {
        if (window.confirm('Toggle Expander 3')) {
          setExpanderThreeOpen(!open);
        }
      }}
    >
      <ExpanderTitleButton>Test expander 3</ExpanderTitleButton>
      <ExpanderContent>Test expander content 3</ExpanderContent>
    </Expander>
  </ExpanderGroup>
</>;
```
