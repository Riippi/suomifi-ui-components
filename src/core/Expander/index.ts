export { Expander, ExpanderProps } from './Expander/Expander';
export {
  ExpanderGroup,
  ExpanderGroupProps,
} from './ExpanderGroup/ExpanderGroup';
export {
  ExpanderContent,
  ExpanderContentProps,
} from './ExpanderContent/ExpanderContent';
export {
  ExpanderTitleButton,
  ExpanderTitleButtonProps,
} from './ExpanderTitle/ExpanderTitleButton';
