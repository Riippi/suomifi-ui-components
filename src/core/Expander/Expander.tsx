import React, { Component, Fragment } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import { baseStyles } from './Expander.baseStyles';
import {
  Expander as CompExpander,
  ExpanderProps as CompExpanderProps,
} from '../../components/Expander/Expander';
import { Icon } from '../Icon/Icon';
import classnames from 'classnames';
import {
  ExpanderGroupProps,
  ExpanderGroup,
  ExpanderGroupConsumer,
} from './ExpanderGroup';

const iconClassName = 'fi-expander_title-icon';
const iconOpenClassName = `${iconClassName}--open`;
const noPaddingClassName = `fi-expander_content--no-padding`;

type ExpanderVariant = 'expander' | 'expanderGroup';

export interface ExpanderProps extends CompExpanderProps, TokensProp {
  /** Remove padding from expandable content area (for background usage with padding in given container etc.) */
  noPadding?: boolean;
  /**
   * 'expander' | 'expanderGroup'
   * @default expander
   */
  variant?: ExpanderVariant;
}

const StyledExpander = styled(
  ({ tokens, noPadding, ...passProps }: ExpanderProps & InternalTokensProp) => {
    return (
      <CompExpander
        {...passProps}
        contentProps={{
          className: classnames({ [noPaddingClassName]: noPadding }),
        }}
      />
    );
  },
)`
  ${(props) => baseStyles(props)};
`;

interface ExpanderState {
  openState: boolean;
}

/**
 * <i class="semantics" />
 * Used for openable expander
 */
class ExpanderItem extends Component<ExpanderProps> {
  static group = (props: ExpanderGroupProps) => {
    return <ExpanderGroup {...withSuomifiDefaultProps(props)} />;
  };

  /** State is only used to update the caret-icon */
  state: ExpanderState = {
    openState:
      this.props.defaultOpen !== undefined ? this.props.defaultOpen : false,
  };

  componentDidUpdate(prevProps: ExpanderProps, prevState: ExpanderState) {
    if (
      !!this.props.consumer &&
      !!prevProps.consumer &&
      this.props.consumer.toggleAllExpanderState !==
        prevProps.consumer.toggleAllExpanderState
    ) {
      const { openState } = this.state;
      const {
        expanderGroup,
        index,
        consumer: { toggleAllExpanderState },
        open,
      } = this.props;

      if (
        !!expanderGroup &&
        index !== undefined &&
        ((open === undefined &&
          !!openState !== toggleAllExpanderState.toState) ||
          (open !== undefined && !!open !== toggleAllExpanderState.toState))
      ) {
        this.handleClick();
      }
    }

    if (
      prevState.openState !== this.state.openState ||
      prevProps.open !== this.props.open
    ) {
      const {
        expanderGroup,
        index,
        consumer: { onClick: consumerOnClick } = { onClick: undefined },
      } = this.props;
      if (!!expanderGroup && !!consumerOnClick && index !== undefined) {
        consumerOnClick(index);
      }
    }
  }

  handleClick = () => {
    const { open, onClick } = this.props;
    const { openState } = this.state;
    const controlled = open !== undefined;
    const newOpenState = controlled ? !!open : !openState;

    this.setState({ openState: newOpenState });
    if (!!onClick) {
      onClick({ openState: newOpenState });
    }
  };

  render() {
    const {
      variant = 'expander',
      open,
      title,
      titleTag,
      index,
      consumer,
      ...passProps
    } = withSuomifiDefaultProps(this.props);
    const openState = open !== undefined ? !!open : this.state.openState;

    if (variant === 'expanderGroup' && 'openAll' in passProps) {
      return <ExpanderGroup {...(passProps as ExpanderGroupProps)} />;
    }

    return (
      <StyledExpander
        {...passProps}
        index={index}
        onClick={this.handleClick}
        open={open}
        titleTag={titleTag}
        title={
          <Fragment>
            {title}
            {!titleTag && (
              <Icon
                icon="chevronDown"
                className={classnames(iconClassName, {
                  [iconOpenClassName]: openState,
                })}
                color={passProps.tokens.colors.highlightBase}
              />
            )}
          </Fragment>
        }
      />
    );
  }
}

export class Expander extends Component<ExpanderProps> {
  static group = (props: ExpanderGroupProps) => {
    return <ExpanderGroup {...withSuomifiDefaultProps(props)} />;
  };

  render() {
    return !!this.props.expanderGroup ? (
      <ExpanderGroupConsumer>
        {(consumer) => <ExpanderItem {...this.props} consumer={consumer} />}
      </ExpanderGroupConsumer>
    ) : (
      <ExpanderItem {...this.props} />
    );
  }
}
