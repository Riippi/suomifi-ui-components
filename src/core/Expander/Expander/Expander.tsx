import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { withSuomifiDefaultProps } from '../../theme/utils';
import { AutoId } from '../../../utils/AutoId';
import { TokensProp, InternalTokensProp } from '../../theme';
import { HtmlDiv } from '../../../reset';
import { baseStyles } from './Expander.baseStyles';
import {
  ExpanderGroupConsumer,
  ExpanderGroupProviderState,
} from '../ExpanderGroup/ExpanderGroup';

const baseClassName = 'fi-expander';
const openClassName = `${baseClassName}--open`;

export interface ExpanderProviderState {
  /** Callback for communicating ExpanderTitle button event to Expander  */
  onToggleExpander: () => void;
  /** Open state for expander */
  open: boolean;
  /** Id for expander button */
  titleId: string | undefined;
  /** Id for expander content */
  contentId: string | undefined;
}

const defaultProviderValue: ExpanderProviderState = {
  onToggleExpander: () => null,
  open: false,
  titleId: undefined,
  contentId: undefined,
};

const {
  Provider: ExpanderProvider,
  Consumer: ExpanderConsumer,
} = React.createContext(defaultProviderValue);

interface InternalExpanderProps {
  /**
   * Children, extend type ExpanderTitleBaseProps or ExpanderContentBaseProps
   * ExpanderProviderState context is used to communicate between title, content and expander
   */
  children: ReactNode;
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Id for expander, must be unique. Duplicate id's break ExpanderGroup functionality.
   * Autogenerated if not provided
   */
  id?: string;
  /** Default status of expander open
   * @default false
   */
  defaultOpen?: boolean;
  /** Controlled open property */
  open?: boolean;
  /** Event handler to execute when clicked */
  onOpenChange?: (open: boolean) => void;
  consumer?: ExpanderGroupProviderState;
}

export interface ExpanderProps extends InternalExpanderProps, TokensProp {}

export interface ExpanderTitleBaseProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Expander consumer for open state and toggle open callback */
  consumer: ExpanderProviderState;
}

export interface ExpanderContentBaseProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Expander consumer for open state */
  consumer: ExpanderProviderState;
}

interface BaseExpanderProps extends InternalExpanderProps {
  id: string;
}

class BaseExpander extends Component<BaseExpanderProps> {
  state: ExpanderState = {
    openState: this.props.defaultOpen || false,
  };

  constructor(props: BaseExpanderProps) {
    super(props);
    if (!!props.consumer) {
      const defaultOpen =
        props.open !== undefined ? props.open : props.defaultOpen || false;
      props.consumer.onExpanderOpenChange(props.id, defaultOpen);
    }
  }

  componentDidUpdate(prevProps: ExpanderProps, prevState: ExpanderState) {
    const { consumer, open } = this.props;
    const controlled = open !== undefined;
    if (
      !!consumer &&
      prevProps.id !== undefined &&
      prevProps.id !== this.props.id
    ) {
      consumer.onExpanderOpenChange(prevProps.id, undefined);
    }
    if (
      !!consumer &&
      consumer.toggleAllExpanderState !==
        prevProps.consumer?.toggleAllExpanderState
    ) {
      if (
        (!controlled &&
          !!this.state.openState !== consumer.toggleAllExpanderState.toState) ||
        (controlled && open !== consumer.toggleAllExpanderState.toState)
      ) {
        this.handleOpenChange();
      }
    }
    if (
      (!controlled && prevState.openState !== this.state.openState) ||
      (controlled && prevProps.open !== open)
    ) {
      if (!!consumer && this.props.id !== undefined) {
        const currentState = controlled ? !!open : this.state.openState;
        consumer.onExpanderOpenChange(this.props.id, currentState);
      }
    }
  }

  componentWillUnmount() {
    if (!!this.props.consumer && !!this.props.consumer.onExpanderOpenChange) {
      this.props.consumer.onExpanderOpenChange(this.props.id, undefined);
    }
  }

  handleOpenChange = () => {
    const { open, onOpenChange } = this.props;
    const { openState } = this.state;
    const controlled = open !== undefined;
    const newOpenState = controlled ? !!open : !openState;
    if (!controlled) {
      this.setState({ openState: newOpenState });
    }
    if (!!onOpenChange) {
      onOpenChange(newOpenState);
    }
  };

  render() {
    const {
      id,
      open,
      defaultOpen,
      onOpenChange,
      className,
      children,
      consumer,
      ...passProps
    } = this.props;
    const openState = open !== undefined ? !!open : this.state.openState;
    const titleId = `${id}_title`;
    const contentId = `${id}_content`;

    return (
      <HtmlDiv
        id={id}
        {...passProps}
        className={classnames(className, baseClassName, {
          [openClassName]: !!openState,
        })}
      >
        <ExpanderProvider
          value={{
            open: openState,
            contentId,
            titleId,
            onToggleExpander: this.handleOpenChange,
          }}
        >
          {children}
        </ExpanderProvider>
      </HtmlDiv>
    );
  }
}

const StyledExpander = styled(
  ({
    tokens,
    id: propId,
    ...passProps
  }: ExpanderProps & InternalTokensProp) => (
    <AutoId id={propId}>
      {(id) => <BaseExpander id={id} {...passProps} />}
    </AutoId>
  ),
)`
  ${(props) => baseStyles(props)};
`;

interface ExpanderState {
  openState: boolean;
}

/**
 * <i class="semantics" />
 * Hide or show content with always visible title
 */
export class Expander extends Component<ExpanderProps> {
  render() {
    return (
      <ExpanderGroupConsumer>
        {(consumer) => (
          <StyledExpander
            {...withSuomifiDefaultProps(this.props)}
            consumer={consumer}
          />
        )}
      </ExpanderGroupConsumer>
    );
  }
}

export { ExpanderConsumer };
