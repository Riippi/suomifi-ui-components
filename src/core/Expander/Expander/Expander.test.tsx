import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../utils/test/axe';
import { Expander, ExpanderProps } from './Expander';
import {
  ExpanderContent,
  ExpanderTitleButton,
  ExpanderTitleButtonProps,
} from '../';

describe('Basic expander', () => {
  const TestExpanderWithProps = (titleProps: ExpanderTitleButtonProps) => {
    return (
      <Expander className="expander-test">
        <ExpanderTitleButton
          {...titleProps}
          {...{ 'data-testid': 'expander-title' }}
        />
        <ExpanderContent>Test expander content</ExpanderContent>
      </Expander>
    );
  };

  const TestExpander = TestExpanderWithProps({
    ariaCloseText: 'click to close expander',
    ariaOpenText: 'click to open expander',
    children: 'Test expander',
  });

  it('render with the same component on the same container does not remount', () => {
    const { getByTestId, rerender } = render(TestExpander);
    expect(getByTestId('expander-title').textContent).toBe(
      'Test expanderclick to open expander',
    );

    // re-render the same component with different props
    rerender(
      TestExpanderWithProps({
        ariaCloseText: 'click to close expander',
        ariaOpenText: 'click to open expander',
        children: 'Test expander two',
      }),
    );
    expect(getByTestId('expander-title').textContent).toBe(
      'Test expander twoclick to open expander',
    );
  });

  it('shoud match snapshot', () => {
    const expanderRenderer = render(TestExpander);
    const { container } = expanderRenderer;
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', axeTest(TestExpander));
});

describe('ExpanderTitleButton Aria attributes', () => {
  const TestExpanderWithProps = (titleProps: ExpanderTitleButtonProps) => {
    return (
      <Expander className="expander-test">
        <ExpanderTitleButton
          {...titleProps}
          {...{ 'data-testid': 'expander-title' }}
        />
        <ExpanderContent>Test expander content</ExpanderContent>
      </Expander>
    );
  };

  it('should be passed to title button', () => {
    const TestExpander = TestExpanderWithProps({
      ariaCloseText: 'click to close expander',
      ariaOpenText: 'click to open expander',
      children: 'Test expander',
    });
    const { getByText, getByRole } = render(TestExpander);
    const button = getByRole('button');
    expect(getByText('click to open expander')).toBeTruthy();
    fireEvent.click(button);
    expect(getByText('click to close expander')).toBeTruthy();
  });
});

describe('Custom id', () => {
  const ExpanderWithCustomId = (
    <Expander id="test-id" {...{ 'data-testid': 'expander-custom-id' }}>
      <ExpanderTitleButton
        toggleButtonProps={{ 'data-testid': 'expander-custom-id-title' }}
      >
        Test expander with custom id
      </ExpanderTitleButton>
      <ExpanderContent {...{ 'data-testid': 'expander-custom-id-content' }}>
        Test expander with custom id content
      </ExpanderContent>
    </Expander>
  );

  it('is passed on to expander', () => {
    const { getByTestId } = render(ExpanderWithCustomId);
    const div = getByTestId('expander-custom-id');
    expect(div).toHaveAttribute('id', 'test-id');
  });

  it('is passed on to title', () => {
    const { getByTestId } = render(ExpanderWithCustomId);
    const div = getByTestId('expander-custom-id-title');
    expect(div).toHaveAttribute('id', 'test-id_title');
  });

  it('is passed on to content', () => {
    const { getByTestId } = render(ExpanderWithCustomId);
    const div = getByTestId('expander-custom-id-content');
    expect(div).toHaveAttribute('id', 'test-id_content');
  });
});

describe('defaultOpen', () => {
  const DefaultOpenExpander = (props?: Omit<ExpanderProps, 'children'>) => (
    <Expander {...props} defaultOpen={true}>
      <ExpanderTitleButton
        {...{ 'data-testid': 'expander-open-by-default-title' }}
      >
        Test expander open by default
      </ExpanderTitleButton>
      <ExpanderContent>Test expander open by default content</ExpanderContent>
    </Expander>
  );

  it('gives the classname to expander title and icon', () => {
    const { getByTestId } = render(DefaultOpenExpander());
    const div = getByTestId('expander-open-by-default-title');
    expect(div).toHaveClass('fi-expander_title-button--open');
    expect(div.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });

  it('classnames will be removed when clicked', () => {
    const mockClickHandler = jest.fn();
    const { getByTestId, getByRole } = render(
      DefaultOpenExpander({ onOpenChange: mockClickHandler }),
    );
    const buttonToClick = getByRole('button');
    const titleDiv = getByTestId('expander-open-by-default-title');
    fireEvent.click(buttonToClick);
    expect(titleDiv).toHaveClass('fi-expander_title-button');
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
    expect(buttonToClick.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });
});

describe('onOpenChange', () => {
  it('is called', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole } = render(
      <Expander onOpenChange={mockClickHandler}>
        <ExpanderTitleButton>Test expander open by default</ExpanderTitleButton>
        <ExpanderContent>Test expander open by default content</ExpanderContent>
      </Expander>,
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
  });
});

describe('open', () => {
  const ControlledExpander = (props?: Partial<ExpanderProps>) => (
    <Expander className="expander-onclick-test" open={true} {...props}>
      <ExpanderTitleButton {...{ 'data-testid': 'expander-title-id' }}>
        Test expander onClick testing
      </ExpanderTitleButton>
      <ExpanderContent>Test expander click testing content</ExpanderContent>
    </Expander>
  );

  it('open-classnames should be found ', async () => {
    const { getByTestId } = render(ControlledExpander());
    const div = getByTestId('expander-title-id');
    expect(div).toHaveClass('fi-expander_title-button--open');

    expect(div.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });

  it('is clicked. Should not change as it is controlled outside', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole, getByTestId } = render(
      ControlledExpander({ onOpenChange: mockClickHandler }),
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    const div = getByTestId('expander-title-id');
    expect(div).toHaveClass('fi-expander_title-button--open');

    expect(button.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });
});
