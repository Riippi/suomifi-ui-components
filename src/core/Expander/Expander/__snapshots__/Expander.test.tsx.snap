// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic expander shoud match snapshot 1`] = `
.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c3:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c3::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c3::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c3::-webkit-inner-spin-button {
  height: auto;
}

.c3::-webkit-outer-spin-button {
  height: auto;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5 {
  position: absolute;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c1 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0,0%,100%);
  position: relative;
  padding: 0;
  border-radius: 2px;
  box-shadow: 0 1px 2px 0 rgba(41,41,41,0.14),0 1px 5px 0 rgba(41,41,41,0.12);
  display: block;
  width: 100%;
  max-width: 100%;
}

.c1:before {
  background-color: hsl(212,63%,98%);
  opacity: 0;
}

.c7 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0,0%,100%);
  border-radius: inherit;
  position: relative;
  visibility: hidden;
  display: block;
  height: 0;
  overflow: hidden;
  word-break: break-word;
  -webkit-transform: scaleY(0);
  -ms-transform: scaleY(0);
  transform: scaleY(0);
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  will-change: transition,height;
}

.c7:not(.fi-expander_content--no-padding) {
  padding: 0 16px;
}

.c7.fi-expander_content--open {
  visibility: visible;
  height: 10%;
  overflow: visible;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
  -webkit-animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
  animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
}

.c7.fi-expander_content--open:not(.fi-expander_content--no-padding) {
  padding-top: 0;
  padding-right: 20px;
  padding-bottom: 20px;
  padding-left: 20px;
}

.c6 {
  display: inline-block;
  vertical-align: baseline;
}

.c2 {
  color: hsl(0,0%,16%);
  position: relative;
  display: block;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  background-color: hsl(212,63%,98%);
  border-radius: inherit;
}

.c2.fi-expander_title-button--open {
  background-color: hsl(0,0%,100%);
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.c2 .fi-expander_title-button_button {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  font-size: font-family:'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(212,63%,45%);
  display: inline-block;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  padding: 17px 60px 16px 20px;
}

.c2 .fi-expander_title-button_button:focus {
  outline: 0;
}

.c2 .fi-expander_title-button_button:focus-within {
  outline: 0;
}

.c2 .fi-expander_title-button_button:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c2 .fi-expander_title-button_button:not(:focus-visible) {
  outline: 0;
}

.c2 .fi-expander_title-button_button:not(:focus-visible):after {
  content: none;
}

.c2 .fi-expander_title-button_button * {
  cursor: pointer;
}

.c2 .fi-expander_title-button_button:hover,
.c2 .fi-expander_title-button_button:active,
.c2 .fi-expander_title-button_button:focus,
.c2 .fi-expander_title-button_button:focus-within {
  cursor: pointer;
}

.c2 .fi-expander_title-button-icon {
  position: absolute;
  top: 0;
  right: 0;
  height: 20px;
  width: 20px;
  margin: 20px;
}

.c2 .fi-expander_title-button--open .fi-expander_title-button-icon,
.c2 .fi-expander_title-button-icon--open {
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

<div
  class="c0 c1 expander-test fi-expander"
  id="4"
>
  <div
    class="c0 c2 fi-expander_title-button"
    data-testid="expander-title"
  >
    <button
      aria-controls="4_content"
      aria-expanded="false"
      class="c3 fi-expander_title-button_button"
      id="4_title"
      type="button"
    >
      Test expander
      <span
        class="c4 c5 fi-visually-hidden"
      >
        click to open expander
      </span>
      <svg
        aria-hidden="true"
        class="fi-icon c6 fi-expander_title-button-icon"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M13 15.6c-.3.3-.6.4-1 .4s-.7-.1-1-.4l-5.6-5.3c-.5-.5-.5-1.4 0-1.9s1.4-.5 2 0l4.6 4.4 4.6-4.4c.5-.5 1.4-.5 2 0 .5.5.5 1.4 0 1.9L13 15.6z"
        />
      </svg>
    </button>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="4_title"
    class="c0 c7 fi-expander_content"
    id="4_content"
    role="region"
  >
    Test expander content
  </div>
</div>
`;
